### ISSUE TITLE
pnpm install --filter installs dependencies for the entire workspace

### ISSUE THREAD

**Issue by bdoof** [pnpm/pnpm#8318]
It is a follow-up for the closed issue --
https://github.com/pnpm/pnpm/issues/6300#issuecomment-2233614366

@zkochan it doesn't work with `shared-workspace-lockfile=true`, as you suggested in your comment above. It still installs dependencies for the entire workspace.

## UPDATE: April 7, 2025 (tl;dr)
You can use this as a workaround:
```
pnpm install --ignore-workspace --frozen-lockfile --lockfile-dir=../
```
Explanation - https://github.com/pnpm/pnpm/issues/8318#issuecomment-2782823861

**Comment by lucashorward** [comment]
I'm having the same issue on 9.5.0, with .nvmrc:
```
20.11.1
```

and .npmrc
```
save-prefix=''
engine-strict=true
shamefully-hoist=true
shared-workspace-lockfile=true
dedupe-peer-dependents=false
```

Packages like `@angular` get installed when they are not dependencies:
<img width="1437" alt="image" src="https://github.com/user-attachments/assets/03c5f810-546f-44bc-b267-28bddf4c1ecb">

**Comment by oieduardorabelo** [comment]
```dockerfile
```

**Comment by oieduardorabelo** [comment]
same issue inside my Docker image too:

```dockerfile
FROM --platform=x86_64 node:20-alpine AS builder

WORKDIR /monorepo

ENV NODE_ENV production
ENV NEXT_TELEMETRY_DISABLED 1

COPY ./package.json ./pnpm-lock.yaml ./pnpm-workspace.yaml ./
COPY ./apps/frontend/package.json ./apps/frontend/package.json
COPY ./apps/frontend-storybook/ ./apps/frontend-storybook/

RUN corepack enable pnpm && \
    pnpm config set dedupe-peer-dependents true && \
    pnpm config set shared-workspace-lockfile true && \
    pnpm config list && \
    pnpm install --prod

COPY ./apps/frontend/ ./apps/frontend/

RUN pnpm --filter @monorepo/frontend build:no-lint
```

and it tries to install the `devDependencies`, which includes a private package:

```bash
#9 [builder 2/8] WORKDIR /monorepo
#9 CACHED

#10 [builder 4/8] COPY ./apps/frontend/package.json ./apps/frontend/package.json
#10 CACHED

#11 [builder 3/8] COPY ./package.json ./pnpm-lock.yaml ./pnpm-workspace.yaml ./
#11 CACHED

#12 [builder 5/8] COPY ./apps/frontend-storybook/ ./apps/frontend-storybook/
#12 CACHED

#13 [builder 6/8] RUN corepack enable pnpm &&     pnpm config set dedupe-peer-dependents true &&     pnpm config set shared-workspace-lockfile true &&     pnpm config list &&     pnpm install --prod
#13 1.181 ! Corepack is about to download https://registry.npmjs.org/pnpm/-/pnpm-9.5.0.tgz
#13 3.478 dedupe-peer-dependents=true
#13 3.478 registry=https://registry.npmjs.org/
#13 3.478 shared-workspace-lockfile=true
#13 3.478 user-agent=pnpm/9.5.0 npm/? node/v20.15.1 linux x64
#13 4.216 Scope: all 3 workspace projects
#13 4.974 Progress: resolved 0, reused 0, downloaded 1, added 0
#13 5.976 Progress: resolved 48, reused 0, downloaded 48, added 0
#13 6.978 Progress: resolved 109, reused 0, downloaded 109, added 0
#13 7.332 /monorepo/apps/frontend:
#13 7.332  ERR_PNPM_FETCH_404  GET https://registry.npmjs.org/@my-org/headless-oauth/-/headless-oauth-3.0.0.tgz: Not Found - 404
#13 7.332 
#13 7.332 This error happened while installing a direct dependency of /monorepo/apps/frontend
#13 7.332 
#13 7.332 No authorization header was set for the request.
```

the `@my-org/headless-oauth` is declared inside my `devDependencies` and should not be part of the Docker build / install

**Comment by zkochan** [comment]
This issue is about `shared-workspace-lockfile=false`. If you don't have this setting on, create a new as this setting is a completely different code path.

**Comment by zkochan** [comment]
I repeat. pnpm performs a completely different workflow, with `shared-workspace-lockfile=true` vs `false`. If you want unrelated issues to be mixed here, it is fine by me.

@lucashorward I could not reproduce your issue. Did you run the command on a clean repository? I don't think that filtered install will remove dependencies if they were previously added with a full install.

@oieduardorabelo your issue is not related to this at all. This issue is about `install --filter` and your install is not using filtering for picking a subset of projects in a workspace. Your issue is with `install --prod`

**Comment by oieduardorabelo** [comment]
oh right, understood. I apologize for mixing it up with this thread! 👍

**Comment by lucashorward** [comment]
@zkochan yes, it happens even when I remove the node_modules folder before running the filtered install.

Here is a repo with reproduction steps in the README.md:

https://github.com/lucashorward/reproduce-pnpm-issue

Please let me know if it's an issue with pnpm or my configuration. The .npmrc and .nvmrc are included in the repo.

**Comment by gabegorelick** [comment]
I am also experiencing this issue. `pnpm install --filter $workspacePackage` installs dependencies for all workspace packages. Toggling `dedupe-peer-dependents` makes no difference. I am also running `shared-workspace-lockfile` with its default (`true`).

Tested with pnpm v9.15.0 on a clean repository with no node_modules.

> This issue is about shared-workspace-lockfile=false. If you don't have this setting, create a new issue.

I'm happy to file a new issue, but by all accounts `shared-workspace-lockfile=false` doesn't appear to be relevant here.

**Comment by maelp** [comment]
I have the same issue

**Comment by louislatreille-cb** [comment]
@bdoof Thanks for the workaround!

Unfortunately, it only works if the package doesn't reference other packages in a monorepo. Else,

```
In : "<package>@workspace:^" is in the dependencies but no package named "<package>" is present in the workspace
```

**Comment by gabegorelick** [comment]
I am still seeing this issue on pnpm v10.11.1.

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm install --filter <project>` installs that project's dependencies only, with a shared lockfile, and leaves the other projects without node_modules.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
